File: AlgorithmsAndDataStructures/Katas/CSharp/getParentOrgs.cs
using System;
using System.Collections.Generic;

namespace getParentOrgs
{
    public class Program
    {
        static void Main(string[] args)
        {
            var orgsList = new List<Org>() 
            {
                new Org { Name = "ANU", ID = 1, ParentID = null },
                new Org { Name = "ANU Law School", ID=2, ParentID = 1 },
                new Org { Name = "College of Law", ID = 3, ParentID = 2 },
            };

            var targetOrg = orgsList[2];
            var parentOrgs = OrgMethods.GetParentOrgsIteratively(orgsList, targetOrg);
            var parentOrgsRecursively = OrgMethods.GetParentOrgsIteratively(orgsList, targetOrg);

            foreach(var org in parentOrgs)
            {
                Console.WriteLine(org.Name);
            }

            foreach (var org in parentOrgsRecursively)
            {
                Console.WriteLine(org.Name);
            }

        }

        public class Org
        {
            public string Name { get; set; }
            public int ID { get; set; }
            public int? ParentID { get; set; }
        }

        public static class OrgMethods
        {
            public static List<Org> GetParentOrgsIteratively(List<Org> orgsList, Org targetOrg)
            {
                var returnList = new List<Org>() { targetOrg };
                if (!targetOrg.ParentID.HasValue) return returnList;

                while (targetOrg.ParentID.HasValue)
                {
                    var parentOrg = orgsList.Find(o => o.ID == targetOrg.ParentID.Value);
                    returnList.Insert(0, parentOrg);
                    targetOrg = parentOrg;
                }

                return returnList;
            }

            public static List<Org> GetParentOrgsRecursively(List<Org> orgsList, Org targetOrg)
            {
                if (!targetOrg.ParentID.HasValue) return new List<Org>() { targetOrg };

                var parentOrg = orgsList.Find(o => o.ID == targetOrg.ParentID);
                return new List<Org> (GetParentOrgsRecursively(orgsList, parentOrg)) { parentOrg };
            }
        }
    }
}

File: AlgorithmsAndDataStructures/Katas/getParentOrgs.js
const orgsList = [
  {
    id: 1,
    name: 'ANU',
    parentId: null,
  },
  {
    id: 2,
    name: 'ANU Law School',
    parentId: 1,
  },
  {
    id: 3,
    name: 'College of Law',
    parentId: 2,
  },
]

const targetOrg = {
  id: 3,
  name: 'College of Law',
  parentId: 2,
}

// recursive version
function getParentOrgs(orgsList, targetOrg) {
  if (targetOrg.parentId === null) return targetOrg
  let newArr = []
  let newTargetOrg = orgsList.find((o) => o.id === targetOrg.parentId)
  return newArr.concat(getParentOrgs(orgsList, newTargetOrg), targetOrg)
}
console.log(getParentOrgs(orgsList, targetOrg))

// iterative version
function getParentsIteratively(orgsList, targetOrg) {
  let returnArray = [targetOrg]
  if (targetOrg.parentId === null) return [targetOrg]

  while (targetOrg.parentId !== null) {
    let parentOrg = orgsList.find((o) => o.id === targetOrg.parentId)
    returnArray.unshift(parentOrg)
    targetOrg = parentOrg
  }
  return returnArray
}

console.log(getParentsIteratively(orgsList, targetOrg))
